{% extends "layout.html" %}
{% block body %}
    <div class="container">
        <h1>About UCLex</h1>
        <div class="row" style="font-size: 16px;">
            <div>
                <p>
                    The UCL Exome (UCLex) Consortium is a coalition of investigators seeking to aggregate and harmonize exome sequencing data from a variety of large-scale sequencing projects, and to make summary data available for the wider scientific community.
                </p>
                <p>
                    The data set provided on this website spans <span class="number_samples"></span> individuals sequenced as part of various disease-specific and population genetic studies. We have removed individuals affected by severe pediatric disease, so this data set should serve as a useful reference set of allele frequencies for severe disease studies. All of the raw data from these projects have been reprocessed through the same pipeline, and jointly variant-called to increase consistency across projects.
                </p>
                <p>
                    A list of the UCLex Principal Investigators and groups that have contributed data to the current release is available below.
                </p>
                <p>
                    All data here are released under a <a href="http://www.wellcome.ac.uk/stellent/groups/corporatesite/@policy_communications/documents/web_document/wtd003207.pdf">Fort Lauderdale Agreement</a> for the benefit of the wider biomedical community. You can freely download and search the data, and use it for publications focused on specific sets of variants (for instance, assessing the frequency of a set of candidate causal variants observed in a collection of rare disease patients). However, we ask that you not publish global (genome-wide) analyses of these data until after the ExAC flagship paper has been published, estimated to be in late 2015. If you’re uncertain which category your analyses fall into, please <a href="mailto:exomeconsortium@gmail.com">email us</a>.
                </p>
                <p>
                    The data are available under the <a href="http://opendatacommons.org/licenses/odbl/1.0/">ODC Open Database License (ODbL)</a> (summary available <a href="http://www.opendatacommons.org/licenses/odbl/1-0/summary/">here</a>): you are free to share and modify the ExAC data so long as you attribute any public use of the database, or works produced from the database; keep the resulting data-sets open; and offer your shared or adapted version of the dataset under the same ODbL license.
                </p>
                <p>
                    The aggregation and release of summary data from the exomes collected by the Exome Aggregation Consortium has been approved by the Partners IRB (protocol 2013P001339, “Large-scale aggregation of human genomic data”).
                </p>
            </div>
            <p>
                To cite the browser, please use: UCLex, London, UK (URL: <a href="http://uclex.cs.ucl.ac.uk">http://uclex.cs.ucl.ac.uk</a>) [date (month, year) accessed].
            </p>
            <p>
                For bug reports, please file an issue on <a href="https://github.com/pontikos/uclex_browser/issues">Github</a>.
            </p>
        </div>
        <div class="row">
            <div class="col-md-3">
                <h4>UCLex Principal Investigators</h4>
                <ul>
                    <li>Tony Segal</li>
                    <li>Andrew Webster</li>
                </ul>
            </div>
            <div class="col-md-4">
                <h4>Contributing projects</h4>
                <ul>
                    <li>1000 Genomes</li>
                    <li>Bulgarian Trios</li>
                    <li>Finland-United States Investigation of NIDDM Genetics (FUSION)</li>
                    <li>GoT2D</li>
                    <li>Inflammatory Bowel Disease</li>
                    <li>METabolic Syndrome In Men (METSIM)</li>
                    <li>Jackson Heart Study</li>
                    <li>Myocardial Infarction Genetics Consortium:</li>
                    <ul>
                        <li>Italian Atherosclerosis, Thrombosis, and Vascular Biology Working Group</li>
                        <li>Ottawa Genomics Heart Study</li>
                        <li>Pakistan Risk of Myocardial Infarction Study (PROMIS)</li>
                        <li>Precocious Coronary Artery Disease Study (PROCARDIS)</li>
                        <li>Registre Gironi del COR (REGICOR)</li>
                    </ul>
                    <li>NHLBI-GO Exome Sequencing Project (ESP)</li>
                    <li>National Institute of Mental Health (NIMH) Controls</li>
                    <li>SIGMA-T2D</li>
                    <li>Sequencing in Suomi (SISu)</li>
                    <li>Swedish Schizophrenia & Bipolar Studies</li>
                    <li>T2D-GENES</li>
                    <li>Schizophrenia Trios from Taiwan</li>
                    <li>The Cancer Genome Atlas (TCGA)</li>
                    <li>Tourette Syndrome Association International Consortium for Genomics (TSAICG)</li>
                </ul>
            </div>
            <div class="col-md-3">
                <h4>Production team</h4>
                <ul>
                    <li>Nikolas Pontikos</li>
                    <li>Vincent Plagnol</li>
                </ul>
                <h4>Analysis team</h4>
                <ul>
                    <li>Nikolas Pontikos</li>
                    <li>Vincent Plagnol</li>
                    <li>Doug Speed</li>
                </ul>
            </div>
            <div class="col-md-2">
                <h4>Website team</h4>
                <ul>
                    <li>Nikolas Pontikos</li>
                </ul>
                <h4>Ethics team</h4>
                <ul>
                    <li></li>
                </ul>
                <h4>UCL Computer Science Cluster</h4>
                <ul>
                    <li>Tristan Clarck</li>
                    <li>David Gregory</li>
                </ul>
                Many thanks to the various clinicians at UCL for generating much of the exome data displayed here and to Tristan Clarck and David Gregory for the compute.
                <br>
                <h4>Funding</h4>
                <ul>
                    <li> (PI: Vincent Plagnol)</li>
                </ul>
            </div>

        </div>
    </div>

{% endblock %}
